sklearn/utils/_testing.py

```diff
@@ -736,111 +736,237 @@ def assert_run_python_script_without_output(source_code, pattern=".+", timeout=6
     dtype : dtype, default=None
-        Force the dtype of the container. Does not apply to `"slice"`
-        container.
+        Force the dtype of the container. Does not apply when `constructor_type` is
+        "slice".
+
+    sparse_container : {"matrix", "array"}, default=None
+        The sparse container to use. Only applies when `constructor_type` is "array"`.
+        Note that if this parameter is not `None` and `container` is 1D of length n,
+        then the converted container will be 2D of shape (1, n).
+
+        - `None` returns a dense numpy array
+        - "matrix" returns a sparse matrix
+        - "array" returns a sparse array and skip test if scipy < 1.8
+
+    sparse_format : {"csc", "csr"}, default="csr"
+        The sparse format to use. Only applies when `constructor_type` is "array" and
+        `sparse_container` is not `None`.
+
+    constructor_lib : {"pandas", "polars", "pyarrow"}, default="pandas"
+        The library to use. Only applies when `constructor_type` is one of "dataframe",
+        "series", and "index". Skip the test if the specified library is not available.
+
+        - "pandas" supports `constructor_type`s "dataframe", "series", and "index"
+        - "polars" supports "dataframe" and "series"
+        - "pyarrow" supports "dataframe" (i.e., Table)
+
     minversion : str, default=None
-        Minimum version for package to install.
+        Minimum version for package to install. Only applies when `constructor_lib` is
+        applicable.
+
+    column_names : index or array-like, default=None
+        The column names of the container. Only applies when `constructor_type` is
+        "dataframe".
+
     categorical_feature_names : list of str, default=None
-        List of column names to cast to categorical dtype.
+        List of column names to cast to categorical dtype. Only applies when
+        `constructor_type` is "dataframe".
 
     Returns
     -------
     converted_container
     """
-    if constructor_name == "list":
-        if dtype is None:
-            return list(container)
-        else:
-            return np.asarray(container, dtype=dtype).tolist()
-    elif constructor_name == "tuple":
-        if dtype is None:
-            return tuple(container)
-        else:
-            return tuple(np.asarray(container, dtype=dtype).tolist())
-    elif constructor_name == "array":
-        return np.asarray(container, dtype=dtype)
-    elif constructor_name in ("pandas", "dataframe"):
+    # Convert the container to the desired dtype (if specified)
+    if sp.sparse.issparse(container):
+        if sparse_container is None:
+            # This means that we are converting from a sparse container to a dense
+            # container, so we densify the container in the first place
+            container = container.toarray()
+        if dtype is not None:
+            container = container.astype(dtype, copy=False)
+    else:
+        container = np.asarray(container, dtype=dtype)
```

Comment:
If I get it right, we are going to convert any dataframe to an array at this stage. I'm not sure this is wise since I could imagine converting a dataframe pandas into a dataframe polars but I would not need to go back to numpy, isn't it?

Reply:
In some way, I would as much as possible delay calling `asarray` to be sure that we actually need it.

Reply:
I didn't think that much of performance given this is a testing utility, but your point is valid. This would mean refactoring and I will do it when I get some full block of time :)

Reply:
> I didn't think that much of performance given this is a testing utility, but your point is valid. 

I'm a bit more scared about the `dtype` casting where will enforce a single `dtype` while we might be interested in dataframe with heterogeneous dtype and just trying to test for pandas and polars. Passing by NumPy in the middle could make it a mess :)

Reply:
@glemaitre I have again refactored the code, and in particular sparse-to-sparse, dataframe-to-dataframe, and series-to-series are dealt with differently. Moreover, now `dtype` should specify the `dtype` of the target (e.g., converting to polars would require a polars dtype) and directly affect the converted container, instead of specifying dtype on the array before conversion (as I previously did).